tripal/tripal
# Set default terms for field properties when chado mapping is missing.

# Bug Fix

### Issue #1939 


### Tripal Version: 4

## Description


Many field properties do not have defaults set for when chado mapping is not present. This causes field addition to fail whenever the column doesn't exist in the chado mapping. Instead this PR add a default for all properties and then if chado mapping provides a more specific term, we use it.

## Testing?



1. Start a new Tripaldocker container
2. Add a type_id to the project table so that we have a column without a chado mapping.
```
docker exec CONTAINERNAME drush sql:query "ALTER TABLE chado.project ADD COLUMN type_id bigint"
docker exec CONTAINERNAME drush sql:query "ALTER TABLE chado.project ADD FOREIGN KEY (type_id) REFERENCES chado.cvterm (cvterm_id) ON DELETE SET NULL;"
```
3. Update the chado schema yaml to ensure Tripal knows about this field. This is needed due to the bug mentioned in issue #1936. Apply the following patch to tripal_chado/chado_schema/chado_schema-1.3.yml.
```diff
diff --git a/tripal_chado/chado_schema/chado_schema-1.3.yml b/tripal_chado/chado_schema/chado_schema-1.3.yml
index 7f482d32f..a73cb9c9d 100644
--- a/tripal_chado/chado_schema/chado_schema-1.3.yml
+++ b/tripal_chado/chado_schema/chado_schema-1.3.yml
@@ -4898,6 +4898,13 @@ project:
     description:
       type: text
       not null: FALSE
+    type_id:
+      type: integer
+  foreign keys:
+    cvterm:
+      table: cvterm
+      columns:
+        type_id: cvterm_id
   unique keys:
     project_c1: name
   primary key: project_id
```
4. Clear the cache using `drush cr`
5. Now in the UI, go to Tripal > Page Structure > Project > Manage Fields in the administrative toolbar.
6. Click on "Create a new field" and add a label of "type", select "Chado Fields", then "Chado Integer Field Type" and click Continue.
7. Thanks to the changes in step 2/3 you now see a "type_id" in the "table column" drop down. Select it and set a term for the field such as "rdfs:type".
8. On 4.x you will now see a WSOD and error message as indicated in the linked issue. On this branch the field will add without error.
9. Now create a project page and confirm the page saves without error and that you can set the type_id :-)
